src/InverseKinematicsIPOPT.cpp

```diff
@@ -0,0 +1,499 @@
+{
+    if(init_z) return false;
+    if(init_lambda) return false;
+
+    Transform p_H_e; //from parent to end effector
+    Eigen::Map< Eigen::VectorXd > x_e (x, 7);
+ 
+    if(init_x){
+        iKDC.setJointPos(desiredJoints);
+        p_H_e = iKDC.getRelativeTransform(parentFrame,endEffectorFrame);
+        x_e.head<3>() = toEigen(p_H_e.getPosition());
+        x_e.tail<4>() = toEigen(p_H_e.getRotation().asQuaternion());
+        
+        for(int i = 7; i < totalDOF; i++){
+            x[i] = desiredJoints(i-7);
+        }
+    }
+    return true;
+}
+
+bool InverseKinematicsIPOPT::eval_f(Ipopt::Index n, const Number* x, bool new_x, Number& obj_value)
+{
+    Eigen::Map< const Eigen::VectorXd > x_in (x, totalDOF);
+    Eigen::Map < Eigen::VectorXd > gradientIn (gradient.data(), totalDOF);
+    Eigen::VectorXd X = x_in;
+    
+    obj_value = 0.5*x_in.transpose()*toEigen(hessian)*x_in;
+    double grad = gradientIn.transpose() * x_in;
+    
+    obj_value += grad;
+
+    return true;
+}
+
+bool InverseKinematicsIPOPT::eval_grad_f(Ipopt::Index n, const Number* x, bool new_x, Number* grad_f)
+{
+    Eigen::Map< const Eigen::VectorXd > x_in (x, totalDOF);
+    Eigen::Map< Eigen::VectorXd > Grad_f (grad_f, totalDOF);
+    Grad_f.setZero();
+    
+    Grad_f = toEigen(hessian)*x_in + toEigen(gradient);
+    
+    return true;
+}
+
+bool InverseKinematicsIPOPT::eval_g(Ipopt::Index n, const Number* x, bool new_x, Ipopt::Index m, Number* g)
+{
+    Transform p_H_e;
+    Eigen::Map< const Eigen::VectorXd > x_in(x, totalDOF);
+    Eigen::Map< Eigen::VectorXd > g_in(g, 8);
+    VectorDynSize joints(totalDOF-7);
+    
+    toEigen(joints) = x_in.tail(totalDOF-7);
+    iKDC.setJointPos(joints);
+    p_H_e = iKDC.getRelativeTransform(parentFrame,endEffectorFrame);
+    
+    g_in.head<3>() = toEigen(p_H_e.getPosition()) - x_in.head<3>();
+    
+    g_in.segment<4>(3) = toEigen(p_H_e.getRotation().asQuaternion()) - x_in.segment<4>(3);
+    
+    return true;
+
+}
+
+bool InverseKinematicsIPOPT::eval_jac_g(Ipopt::Index n, const Number* x, bool new_x, Ipopt::Index m, Ipopt::Index nele_jac, Ipopt::Index* iRow, Ipopt::Index* jCol, Number* values)
+{
+    
+    if(values == NULL){
+        int val = 0;
+        int i = 0;
+        int j = 0;
+    
+        //sparsity of the jacobian
+        for(val = 0; val < 7; val++){
+            iRow[val] = val; jCol[val] = val; //the top left corner of the matrix is diagonal
+        }
+        
+        for(i = 0; i < 7; i++){
+            for(j = 7; j < (totalDOF); j++){
+
+                iRow[val] = i; jCol[val] = j;  //the right top part is dense
+                val++;
+            }
+        }
+
+    }
+    else{
+        Eigen::Map< const Eigen::VectorXd > x_in(x, totalDOF);
+        Eigen::Matrix<double,Eigen::Dynamic,Eigen::Dynamic,Eigen::RowMajor> denseJac(m,totalDOF);
+        Vector4 quaternion;
+        VectorDynSize configuration;
+        MatrixFixSize<7,6> map;
+        MatrixDynSize jacobian;
```

Comment:
The fact that the declaration is in an ``if`` statement, does it cause problems?